Fix yaml request payload send into training method way which lead to unnecessarily retraining NLU model if only modified responses utterances in domain.

liuyuweitarek opened the pull request:
**Proposed changes**:
- Fix yaml request payload send into training method way which lead to unnecessarily retraining NLU model if only modified responses utterances in domain.

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=11576) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=11576) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=11576) it.</sub>

amn41 commented:
thanks for the PR, @liuyuweitarek ! 
I think there might be a piece of missing logic (tbf I haven't done a proper review yet). 

If there is a response selector with `use_text_as_label` set to True, then the responses should be part of the fingerprint. You can check the migration notes here for more detail https://rasa.com/docs/rasa/2.x/migration-guide/#response-selectors

liuyuweitarek commented:
> thanks for the PR, @liuyuweitarek ! I think there might be a piece of missing logic (tbf I haven't done a proper review yet).
> 
> If there is a response selector with `use_text_as_label` set to True, then the responses should be part of the fingerprint. You can check the migration notes here for more detail https://rasa.com/docs/rasa/2.x/migration-guide/#response-selectors

Thanks for your reply, @amn41 ！Yes, you are right!
After further researching, I found my real problem is it wrongly judges to retrain the NLU/CORE model while using the API:"model/train" belong to Rasa API server. Some slight changes, changing an utterance under responses (not belong to response selector), will retrain the NLU model.  

The reason of this problem is because it put the whole request payload into the "single file" and send to the training method.

```python

[ In rasa/server.py 1561~1584 ]
def _training_payload_from_yaml(
    request: Request, temp_dir: Path, file_name: Text = "data.yml"
) -> Dict[Text, Any]:
    logger.debug("Extracting YAML training data from request body.")

    decoded = request.body.decode(rasa.shared.utils.io.DEFAULT_ENCODING)
    _validate_yaml_training_payload(decoded)

    training_data = temp_dir / file_name
    rasa.shared.utils.io.write_text_file(decoded, training_data)              # Here write the whole payload into single file
    
    ...
    return dict(
        domain=str(training_data),             # Training data be put in single file
        config=str(training_data),
        training_files=str(temp_dir),
        output=model_output_directory,
        force_training=rasa.utils.endpoints.bool_arg(request, "force_training", False),
        core_additional_arguments=_extract_core_additional_arguments(request),
        nlu_additional_arguments=_extract_nlu_additional_arguments(request),
    )

[ Send to train ]
training_result = await train_async(**training_payload)
```

In the latest commit, I separated  request payload into several files with standard architecture in the temp folder and made training API work properly as "rasa train" in command. 

I use the data in [08-response-selection](https://github.com/RasaHQ/rasa-2.x-form-examples/tree/main/08-response-selection) to test training api:

1. It will retrain NLU model as we modified the utterance related to response selector.
2. It will only update the model  response without retraining Core/NLU models as we modified the simple utterance unrelated to response selector.

Here is the [Postman config](https://gist.github.com/liuyuweitarek/be83e3429e06d2de386110a2e04440aa) for the old /new request format.

Thanks for your time!

ancalita reviewed (commented):
Thank you for your contribution 💫  I left a few refactoring suggestions 🙏🏻 
Also could you please add a few unit tests in `test_server.py` testing the scenarios you mentioned?
You will also have to resolve the merge conflicts in order for us to run the CI checks.

liuyuweitarek commented on rasa/server.py:
```diff
@@ -1557,33 +1569,58 @@ def _validate_json_training_payload(rjs: Dict) -> None:
+        )
+    
+    if "nlu" in request_payload:
+        nlu_path = os.path.join(temp_dir, API_TEMP_DATA_FOLDER, API_TEMP_NLU_YAML_FILE)
+        nlu_data = {'nlu': request_payload["nlu"]}
+        rasa.shared.utils.io.write_yaml(nlu_data, nlu_path)
```
Thank you @ancalita ! Great suggestion! 
P.S. The reason that wrote these lines repeatedly in the beginning is because of taking each 5 keys as independent processes and not sure whether needing to deal with subkeys under them respectively. With above assumption, easily debugging in each process under the single request out of loop is also be taken into consideration. Whatever, it's great news that there's no need further processing~ 😄!

liuyuweitarek commented on rasa/server.py:
```diff
@@ -1557,33 +1569,58 @@ def _validate_json_training_payload(rjs: Dict) -> None:
 
+    if "domain" in request_payload:
+        domain_path = os.path.join(temp_dir, API_TEMP_DOMAIN_YAML_FILE)
+        domain_data = request_payload["domain"]
```
Not sure with handling all keys to right files and to maintain clean request payload structure, I place certain keys into specific files and form standard training data structure, i.e. 
```yaml
nlu:
  - ...
rules:
  -...
stories:
  -...
config:
  key_config_1
    -...
  key_config_2
    -...
domain:
  key_domain_1
    -...
  key_domain_2
    -...
``` 
rather than

```yaml
nlu:
  - ...
rules:
  -...
stories:
  -...
key_config_1
  -...
key_config_2
  -...
key_domain_1
  -...
key_domain_2
  -...
```

ancalita reviewed (commented):
@liuyuweitarek Apologies for my late reply to your latest changes!
I only noticed that your PR is targeting `2.8.x`, however we don't maintain this major series anymore (please see this [forum post](https://forum.rasa.com/t/what-our-product-release-and-maintenance-policy-means-to-the-community/56096) for more info). 

It would be 💯  amazing if you could open a PR with the most updated PR changes targeting `3.3.x` release branch, which is the latest version we maintain.

liuyuweitarek commented:
@ancalita It's ok. I'm also too busy to add test case on this PR. 😭 
I'll keep track on `3.3.x` and check if it also needs this fixation. Thanks for your time!
Please feel free to close this PR.
